File: ui/src/api/endpoints/quality.ts
/**
 * Quality API Endpoints
 */
import apiClient from '../client';

export interface DefectDetail {
  id: string;
  tenant_id: string;
  defect_timestamp: string;
  detection_point: 'spi' | 'aoi' | 'xray' | 'ict' | 'fct' | 'visual' | 'customer';
  equipment_code: string;
  line_code: string;
  production_order_no?: string;
  product_code: string;
  lot_no?: string;
  panel_id?: string;
  pcb_serial?: string;
  defect_category: string;
  defect_code: string;
  defect_description?: string;
  defect_location?: string;
  component_ref?: string;
  component_code?: string;
  x_position?: number;
  y_position?: number;
  defect_qty: number;
  severity?: 'critical' | 'major' | 'minor';
  image_url?: string;
  detected_by?: string;
  detection_method?: string;
  repair_action?: string;
  repair_result?: 'repaired' | 'scrapped' | 'pending' | 'no_action';
  repaired_by?: string;
  repaired_at?: string;
  root_cause_category?: string;
  root_cause_detail?: string;
  created_at: string;
}

export interface DefectAnalysis {
  defect_code: string;
  defect_name?: string;
  defect_category: string;
  count: number;
  percentage: number;
  cumulative_percentage: number;
}

export interface DefectTrend {
  date: string;
  line_code: string;
  defect_count: number;
  defect_rate: number;
}

export interface InspectionResult {
  id: string;
  tenant_id: string;
  inspection_no: string;
  inspection_type: 'SPI' | 'AOI' | 'AXI' | 'MANUAL' | 'ICT' | 'FCT';
  production_order_id?: string;
  lot_no: string;
  board_id?: string;
  product_code: string;
  line_code?: string;
  equipment_id?: string;
  operation_no?: number;
  inspection_datetime: string;
  shift?: string;
  inspector_code?: string;
  result: 'PASS' | 'FAIL' | 'REWORK';
  total_inspected: number;
  pass_qty: number;
  fail_qty: number;
  defect_points?: Array<{
    location: string;
    defect_code: string;
    x?: number;
    y?: number;
  }>;
  inspection_time_sec?: number;
  rework_flag: boolean;
  created_at: string;
  pass_rate?: number;
}

export interface SPCData {
  id: string;
  tenant_id: string;
  spc_no: string;
  line_code: string;
  equipment_id?: string;
  product_code: string;
  lot_no?: string;
  measurement_type: string;
  measurement_datetime: string;
  sample_size: number;
  measured_value: number;
  unit?: string;
  spec_lower?: number;
  spec_upper?: number;
  target_value?: number;
  control_lower?: number;
  control_upper?: number;
  cpk_value?: number;
  out_of_spec: boolean;
  out_of_control: boolean;
  shift?: string;
  created_at: string;
}

export interface SPCChartData {
  measurement_type: string;
  measurements: SPCData[];
  mean: number;
  std_dev: number;
  ucl: number;
  lcl: number;
  usl?: number;
  lsl?: number;
  cpk?: number;
}

export interface QualitySummary {
  date: string;
  line_code: string;
  total_inspected: number;
  total_pass: number;
  total_fail: number;
  pass_rate: number;
  defect_rate: number;
  top_defects: DefectAnalysis[];
}

export interface TraceabilityData {
  lot_no: string;
  product_code?: string;
  production_order_no?: string;
  status?: string;
  operations: Array<{
    operation_no: number;
    operation_name?: string;
    equipment_code: string;
    line_code: string;
    timestamp: string;
    operator_code?: string;
    material_lots?: Record<string, any>;
    process_parameters?: Record<string, any>;
    quality_results?: Record<string, any>;
  }>;
  material_traceability?: Record<string, any>;
}

// API Functions

export const getInspections = async (params?: {
  page?: number;
  page_size?: number;
  inspection_type?: string;
  lot_no?: string;
  product_code?: string;
  result?: string;
  start_date?: string;
  end_date?: string;
}): Promise<InspectionResult[]> => {
  const response = await apiClient.get('/mes/quality/inspections', { params });
  return response.data;
};

export const createInspection = async (data: Partial<InspectionResult>): Promise<InspectionResult> => {
  const response = await apiClient.post('/mes/quality/inspections', data);
  return response.data;
};

export const getInspection = async (inspectionId: string): Promise<InspectionResult> => {
  const response = await apiClient.get(`/mes/quality/inspections/${inspectionId}`);
  return response.data;
};

export const getDefects = async (params?: {
  page?: number;
  page_size?: number;
  line_code?: string;
  product_code?: string;
  defect_category?: string;
  defect_code?: string;
  start_date?: string;
  end_date?: string;
}): Promise<DefectDetail[]> => {
  const response = await apiClient.get('/mes/quality/defects', { params });
  return response.data;
};

export const createDefect = async (data: Partial<DefectDetail>): Promise<DefectDetail> => {
  const response = await apiClient.post('/mes/quality/defects', data);
  return response.data;
};

export const getDefectAnalysis = async (params?: {
  line_code?: string;
  product_code?: string;
  start_date?: string;
  end_date?: string;
  limit?: number;
}): Promise<DefectAnalysis[]> => {
  const response = await apiClient.get('/mes/quality/defects/analysis', { params });
  return response.data;
};

export const getDefectPareto = async (params?: {
  line_code?: string;
  product_code?: string;
  start_date?: string;
  end_date?: string;
}): Promise<DefectAnalysis[]> => {
  const response = await apiClient.get('/mes/quality/defects/pareto', { params });
  return response.data;
};

export const getSPCData = async (params?: {
  line_code?: string;
  product_code?: string;
  measurement_type?: string;
  start_date?: string;
  end_date?: string;
  page?: number;
  page_size?: number;
}): Promise<SPCData[]> => {
  const response = await apiClient.get('/mes/quality/spc/data', { params });
  return response.data;
};

export const getSPCChartData = async (
  measurementType: string,
  params?: {
    line_code?: string;
    product_code?: string;
    limit?: number;
  }
): Promise<SPCChartData> => {
  const response = await apiClient.get(`/mes/quality/spc/chart/${measurementType}`, { params });
  return response.data;
};

export const getTraceability = async (lotNo: string): Promise<TraceabilityData> => {
  const response = await apiClient.get(`/mes/quality/traceability/${lotNo}`);
  return response.data;
};

File: api/schemas/mes/quality.py
"""
Quality-related Pydantic schemas
"""
from datetime import datetime, date
from decimal import Decimal
from enum import Enum
from typing import Optional, List, Dict, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DetectionPoint(str, Enum):
    """Defect detection points"""
    SPI = "spi"
    AOI = "aoi"
    XRAY = "xray"
    ICT = "ict"
    FCT = "fct"
    VISUAL = "visual"
    CUSTOMER = "customer"


class DefectCategory(str, Enum):
    """Defect categories"""
    SOLDER = "solder"
    COMPONENT = "component"
    PLACEMENT = "placement"
    SHORT = "short"
    OPEN = "open"
    BRIDGE = "bridge"
    INSUFFICIENT = "insufficient"
    VOID = "void"
    CRACK = "crack"
    CONTAMINATION = "contamination"
    MECHANICAL = "mechanical"
    OTHER = "other"


class DefectSeverity(str, Enum):
    """Defect severity levels"""
    CRITICAL = "critical"
    MAJOR = "major"
    MINOR = "minor"


class RepairResult(str, Enum):
    """Repair result values"""
    REPAIRED = "repaired"
    SCRAPPED = "scrapped"
    PENDING = "pending"
    NO_ACTION = "no_action"


class InspectionType(str, Enum):
    """Inspection types"""
    SPI = "SPI"
    AOI = "AOI"
    AXI = "AXI"
    MANUAL = "MANUAL"
    ICT = "ICT"
    FCT = "FCT"


class InspectionResult(str, Enum):
    """Inspection results"""
    PASS = "PASS"
    FAIL = "FAIL"
    REWORK = "REWORK"


# ==================== Defect Type ====================

class DefectTypeResponse(BaseModel):
    """Response schema for defect type"""
    id: UUID
    tenant_id: UUID
    defect_code: str
    defect_name: str
    defect_name_en: Optional[str] = None
    defect_category: str
    severity: DefectSeverity
    detection_methods: Optional[List[str]] = None
    description: Optional[str] = None
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ==================== Defect Detail ====================

class DefectDetailCreate(BaseModel):
    """Schema for creating defect detail"""
    detection_point: DetectionPoint
    equipment_code: str = Field(..., max_length=30)
    line_code: str = Field(..., max_length=20)
    production_order_no: Optional[str] = Field(None, max_length=20)
    product_code: str = Field(..., max_length=30)
    lot_no: Optional[str] = Field(None, max_length=50)
    panel_id: Optional[str] = Field(None, max_length=50)
    pcb_serial: Optional[str] = Field(None, max_length=50)
    defect_category: DefectCategory
    defect_code: str = Field(..., max_length=20)
    defect_description: Optional[str] = None
    defect_location: Optional[str] = Field(None, max_length=100)
    component_ref: Optional[str] = Field(None, max_length=50)
    component_code: Optional[str] = Field(None, max_length=30)
    x_position: Optional[Decimal] = None
    y_position: Optional[Decimal] = None
    defect_qty: int = Field(default=1, ge=1)
    severity: Optional[DefectSeverity] = None
    image_url: Optional[str] = None
    detected_by: Optional[str] = Field(None, max_length=20)
    detection_method: Optional[str] = Field(None, max_length=30)


class DefectDetailUpdate(BaseModel):
    """Schema for updating defect detail"""
    repair_action: Optional[str] = Field(None, max_length=100)
    repair_result: Optional[RepairResult] = None
    repaired_by: Optional[str] = Field(None, max_length=20)
    root_cause_category: Optional[str] = Field(None, max_length=30)
    root_cause_detail: Optional[str] = None


class DefectDetailResponse(BaseModel):
    """Response schema for defect detail"""
    id: UUID
    tenant_id: UUID
    defect_timestamp: datetime
    detection_point: DetectionPoint
    equipment_code: str
    line_code: str
    production_order_no: Optional[str] = None
    product_code: str
    lot_no: Optional[str] = None
    panel_id: Optional[str] = None
    pcb_serial: Optional[str] = None
    defect_category: DefectCategory
    defect_code: str
    defect_description: Optional[str] = None
    defect_location: Optional[str] = None
    component_ref: Optional[str] = None
    component_code: Optional[str] = None
    x_position: Optional[Decimal] = None
    y_position: Optional[Decimal] = None
    defect_qty: int
    severity: Optional[DefectSeverity] = None
    image_url: Optional[str] = None
    detected_by: Optional[str] = None
    detection_method: Optional[str] = None
    repair_action: Optional[str] = None
    repair_result: Optional[RepairResult] = None
    repaired_by: Optional[str] = None
    repaired_at: Optional[datetime] = None
    root_cause_category: Optional[str] = None
    root_cause_detail: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DefectAnalysisResponse(BaseModel):
    """Defect analysis/pareto response"""
    defect_code: str
    defect_name: Optional[str] = None
    defect_category: DefectCategory
    count: int
    percentage: float
    cumulative_percentage: float


class DefectTrend(BaseModel):
    """Defect trend data point"""
    date: date
    line_code: str
    defect_count: int
    defect_rate: float


# ==================== Inspection Result ====================

class InspectionResultCreate(BaseModel):
    """Schema for creating inspection result"""
    inspection_type: InspectionType
    lot_no: str = Field(..., max_length=30)
    board_id: Optional[str] = Field(None, max_length=50)
    product_code: str = Field(..., max_length=30)
    line_code: Optional[str] = Field(None, max_length=20)
    operation_no: Optional[int] = None
    shift: Optional[str] = Field(None, max_length=5)
    inspector_code: Optional[str] = Field(None, max_length=20)
    result: InspectionResult
    total_inspected: int = Field(default=1, ge=1)
    pass_qty: int = Field(default=0, ge=0)
    fail_qty: int = Field(default=0, ge=0)
    defect_points: Optional[List[Dict[str, Any]]] = None
    inspection_time_sec: Optional[Decimal] = None


class InspectionResultResponse(BaseModel):
    """Response schema for inspection result"""
    id: UUID
    tenant_id: UUID
    inspection_no: str
    inspection_type: InspectionType
    production_order_id: Optional[UUID] = None
    lot_no: str
    board_id: Optional[str] = None
    product_code: str
    line_code: Optional[str] = None
    equipment_id: Optional[UUID] = None
    operation_no: Optional[int] = None
    inspection_datetime: datetime
    shift: Optional[str] = None
    inspector_code: Optional[str] = None
    result: InspectionResult
    total_inspected: int
    pass_qty: int
    fail_qty: int
    defect_points: Optional[List[Dict[str, Any]]] = None
    inspection_time_sec: Optional[Decimal] = None
    rework_flag: bool
    created_at: datetime

    # Calculated
    pass_rate: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


# ==================== SPC Data ====================

class SPCDataCreate(BaseModel):
    """Schema for creating SPC data"""
    line_code: str = Field(..., max_length=20)
    product_code: str = Field(..., max_length=30)
    lot_no: Optional[str] = Field(None, max_length=30)
    measurement_type: str = Field(..., max_length=30)
    sample_size: int = Field(default=1, ge=1)
    measured_value: Decimal
    unit: Optional[str] = Field(None, max_length=10)
    spec_lower: Optional[Decimal] = None
    spec_upper: Optional[Decimal] = None
    target_value: Optional[Decimal] = None
    control_lower: Optional[Decimal] = None
    control_upper: Optional[Decimal] = None
    shift: Optional[str] = Field(None, max_length=5)


class SPCDataResponse(BaseModel):
    """Response schema for SPC data"""
    id: UUID
    tenant_id: UUID
    spc_no: str
    line_code: str
    equipment_id: Optional[UUID] = None
    product_code: str
    lot_no: Optional[str] = None
    measurement_type: str
    measurement_datetime: datetime
    sample_size: int
    measured_value: Decimal
    unit: Optional[str] = None
    spec_lower: Optional[Decimal] = None
    spec_upper: Optional[Decimal] = None
    target_value: Optional[Decimal] = None
    control_lower: Optional[Decimal] = None
    control_upper: Optional[Decimal] = None
    cpk_value: Optional[Decimal] = None
    out_of_spec: bool
    out_of_control: bool
    shift: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SPCChartData(BaseModel):
    """SPC chart data"""
    measurement_type: str
    measurements: List[SPCDataResponse]
    mean: float
    std_dev: float
    ucl: float
    lcl: float
    usl: Optional[float] = None
    lsl: Optional[float] = None
    cpk: Optional[float] = None


# ==================== Quality Summary ====================

class QualitySummary(BaseModel):
    """Daily quality summary"""
    date: date
    line_code: str
    total_inspected: int
    total_pass: int
    total_fail: int
    pass_rate: float
    defect_rate: float
    top_defects: List[DefectAnalysisResponse]
